Reference for stage 3 is-error (#38251)
mdn/content: files/en-us/web/javascript/reference/global_objects/error/iserror/index.md

---
title: Error.isError()
slug: Web/JavaScript/Reference/Global_Objects/Error/isError
page-type: javascript-static-method
browser-compat: javascript.builtins.Error.isError
---

{{JSRef}}

The **`Error.isError()`** static method determines whether the passed value is an {{jsxref("Error")}}.

## Syntax

```js-nolint
Error.isError(value)
```

### Parameters

- `value`
  - : The value to be checked.

### Return value

`true` if `value` is an {{jsxref("Error")}}; otherwise, `false`.

## Description

> `Error.isError()` checks if the passed value is an {{jsxref("Error")}}. It uses the same mechanism as {{jsxref("Array.isArray()")}}: it performs a _branded check_, similar to the [`in`](/en-US/docs/Web/JavaScript/Reference/Operators/in) operator, for a private property initialized by the {{jsxref("Error/Error", "Error()")}} constructor. It is a more robust alternative to [`instanceof Error`](/en-US/docs/Web/JavaScript/Reference/Operators/instanceof) because it avoids false positives and false negatives:
> 
> - It does not check the value's prototype chain—`Error.isError()` rejects objects with `Error.prototype` in its prototype chain but aren't actual errors, which `instanceof Error` would accept.
> - For `Error` objects constructed in another realm, the identity of the `Error` constructor is different and would therefore cause `instanceof Error` to return `false` in this realm.

Reviewer: This isn't wrong, but I think this split makes it a bit more readable to put in terms of what isError does, rather than what `instanceof` does.

```suggestion
`Error.isError()` checks if the passed value is an {{jsxref("Error")}}. It uses the same mechanism as {{jsxref("Array.isArray()")}}: it performs a _branded check_, similar to the [`in`](/en-US/docs/Web/JavaScript/Reference/Operators/in) operator, for a private property initialized by the {{jsxref("Error/Error", "Error()")}} constructor. 

It is a more robust alternative to [`instanceof Error`](/en-US/docs/Web/JavaScript/Reference/Operators/instanceof) because it avoids false positives and false negatives:

- `Error.isError()` rejects values that aren't actual `Error` instances, even if they have `Error.prototype` their prototype chain — `instanceof Error` would accept these as it does check the prototype chain.
- `Error.isError()` accepts `Error` objects constructed in another realm — `instanceof Error` returns `false` for these because the identity of the `Error` constructor is different across realms.
```

> Although {{domxref("DOMException")}} is not specified as a real subclass of `Error` (the `Error` constructor is not the prototype of the `DOMException` constructor), `DOMException` still behaves like `Error` for all branded checking purposes, so `Error.isError()` returns `true` for `DOMException` instances.

Reviewer: Suggest we list the "takeaway point" first.

```suggestion
`Error.isError()` returns `true` for `DOMException` instances. This is because, although {{domxref("DOMException")}} is not specified as a real subclass of `Error` (the `Error` constructor is not the prototype of the `DOMException` constructor), `DOMException` still behaves like `Error` for all branded checking purposes.
```

---
title: Error.isError()
slug: Web/JavaScript/Reference/Global_Objects/Error/isError
page-type: javascript-static-method
browser-compat: javascript.builtins.Error.isError
---

{{JSRef}}

The **`Error.isError()`** static method determines whether the passed value is an {{jsxref("Error")}}.

## Syntax

```js-nolint
Error.isError(value)
```

### Parameters

- `value`
  - : The value to be checked.

### Return value

`true` if `value` is an {{jsxref("Error")}}; otherwise, `false`.

## Description

> `Error.isError()` checks if the passed value is an {{jsxref("Error")}}. It uses the same mechanism as {{jsxref("Array.isArray()")}}: it performs a _branded check_, similar to the [`in`](/en-US/docs/Web/JavaScript/Reference/Operators/in) operator, for a private property initialized by the {{jsxref("Error/Error", "Error()")}} constructor.

Reviewer: I think this is what you are saying?

```suggestion
`Error.isError()` checks if the passed value is an {{jsxref("Error")}} by performing a _branded check_ for a private property initialized by the {{jsxref("Error/Error", "Error()")}} constructor.
This is the same mechanism used by {{jsxref("Array.isArray()")}}, which is in turn similar to the mechanism used by the [`in`](/en-US/docs/Web/JavaScript/Reference/Operators/in) operator.
```

So how does this work for `DOMException` ? Its constructor includes the same kind of private property as the error constructor?

Reviewer: Also, possibly
```suggestion
`Error.isError()` checks if the passed value is an {{jsxref("Error")}} or `DOMException`. It does so by performing a _branded check_ for a private property initialized by the {{jsxref("Error/Error", "Error()")}} constructor.
This is the same mechanism used by {{jsxref("Array.isArray()")}}, which is in turn similar to the mechanism used by the [`in`](/en-US/docs/Web/JavaScript/Reference/Operators/in) operator.
```

Author: `DOMException` is _supposed_ to behave like `Error` so I wouldn't want to imply that they are unrelated things and `isError` happens to check both. I'm not sure why the `in` operator reference is removed though—without it it's still hard to conceptualize what "branded check" and "private property" means.

Reviewer: Both suggestions still mentions `in`, it's just that this moved to the third sentence. The problem I am trying to fix is that it is hard to parse your sentence. 

I'm fine with not mentioning "differentiating" the error types, so ignore my second suggestion and let's just consider the first one, which I will reproduce here:

```suggestion
`Error.isError()` checks if the passed value is an {{jsxref("Error")}}. It does so by performing a _branded check_ for a private property initialized by the {{jsxref("Error/Error", "Error()")}} constructor.
This is the same mechanism used by {{jsxref("Array.isArray()")}}, which is in turn similar to the mechanism used by the [`in`](/en-US/docs/Web/JavaScript/Reference/Operators/in) operator.
```

And I am still interested in 

> So how does this work for `DOMException` ? Its constructor includes the same kind of private property as the error constructor?

Author: > So how does this work for `DOMException` ? Its constructor includes the same kind of private property as the error constructor?

I think I replied in one of the resolved conversations above but yeah that's exactly what happens.

Reviewer: If you did, I missed it. Sorry. Thanks.

It is a more robust alternative to [`instanceof Error`](/en-US/docs/Web/JavaScript/Reference/Operators/instanceof) because it avoids false positives and false negatives:

- `Error.isError()` rejects values that aren't actual `Error` instances, even if they have `Error.prototype` their prototype chain — `instanceof Error` would accept these as it does check the prototype chain.
- `Error.isError()` accepts `Error` objects constructed in another realm — `instanceof Error` returns `false` for these because the identity of the `Error` constructor is different across realms.

`Error.isError()` returns `true` for {{domxref("DOMException")}} instances. This is because, although `DOMException` is not specified as a real subclass of `Error` (the `Error` constructor is not the prototype of the `DOMException` constructor), `DOMException` still behaves like `Error` for all branded checking purposes.

## Examples

### Using Error.isError()

```js
// all following calls return true
Error.isError(new Error());
Error.isError(new TypeError());
Error.isError(new DOMException());
try {
  1 + 1n;
} catch (e) {

>   console.log(Error.isError(e)); // Throws a TypeError

Author: ```suggestion
  console.log(Error.isError(e)); // The operation threw a TypeError, so this returns true
```
